sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala

```diff
@@ -119,6 +120,18 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
         limitSQL = limit.sql
       } yield s"$childSQL LIMIT $limitSQL"
 
+    // TABLESAMPLE is part of tableSource clause in the parser,
+    // and thus we must handle it with subquery.
+    case Sample(lb, ub, withReplacement, _, child @ Subquery(alias, grandChild))
```

Comment:
this seems pretty brittle. i wonder if we can change sample to make it very clear this is either a table sample, or a normal sample.